Implement string-valued text-overflow in Servo

Part of #13702

In #13701, we added support for parsing the full range of values that `text-overflow` supports. However, [this is disabled in Servo mode](https://github.com/Manishearth/servo/blob/ea765688cbc7a7f5e223cf060f2bf1341dfa0ae5/components/style/properties/longhand/text.mako.rs#L15)

While supporting the two-value form of text-overflow (e.g. `text-overflow: "..." clip`) is cumbersome, it should be pretty straightforward to support the one-value form.

Steps:
- Remove the conditional in https://github.com/Manishearth/servo/blob/ea765688cbc7a7f5e223cf060f2bf1341dfa0ae5/components/style/properties/longhand/text.mako.rs#L15 , and use the `<%helpers:longhand ...` one unconditionally.
- Try compiling it. Layout will probably fail in [`push_fragment_to_line()` in `components/layout/inline.rs`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/inline.rs#L691), because `text_overflow::clip`/`text_overflow::ellipsis` are now `text_overflow::Side::Clip`/`text_overflow::Side::Ellipsis`, and what used to be an enum is now a struct. Match on `fragment.style().get_text().text_overflow.first` instead, and use  the new names for the variants.
- Make it possible to pass arbitrary strings to [`transform_into_ellipsis()`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/fragment.rs#L1013) as an argument. In `push_fragment_to_line()` add an extra match arm that is similar to the ellipsis one, but pushes a string. In the ellipsis one, pass "…".
- Test that the [lorem ipsum example on MDN works](https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow)

Relevant layout code: [`components/layout/inline.rs`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/inline.rs#L691) , [`components/layout/fragment.rs`](https://github.com/servo/servo/blob/b9bd15ebb0e84b48d5b0e7d1e952a911960d4fef/components/layout/fragment.rs#L1013)

Tests: None, but we will add some when this is done.

`https://github.com/cynicaldevil/servo/commit/74c89e61eece0df4d19ba76996de4ce0238e08bc`
Error:
```
error[E0433]: failed to resolve. Could not find `text_overflow` in `style::properties`
   --> /home/nikhil/workspace/mozilla/servo/components/layout/inline.rs:692:13
    |
692 |             properties::text_overflow::Side::Clip => {}
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `text_overflow` in `style::properties`

error[E0433]: failed to resolve. Could not find `text_overflow` in `style::properties`
   --> /home/nikhil/workspace/mozilla/servo/components/layout/inline.rs:693:13
    |
693 |             properties::text_overflow::Side::Ellipsis => {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `text_overflow` in `style::properties`
```

```
error[E0505]: cannot move out of `fragment` because it is borrowed
    |
691 |         let passed_str: Option<&str> = match fragment.style().get_text().text_overflow.first {
    |                                              -------- borrow of `fragment` occurs here
...
704 |             self.push_fragment_to_line_ignoring_text_overflow(fragment, layout_context);
    |                                                               ^^^^^^^^ move out of `fragment` occurs here

```

--- a/components/layout/fragment.rs
+++ b/components/layout/fragment.rs
@@ -1029,12 +1029,15 @@ impl Fragment {
     }
 
     /// Transforms this fragment into an ellipsis fragment, preserving all the other data.
-    pub fn transform_into_ellipsis(&self, layout_context: &LayoutContext) -> Fragment {
+    pub fn transform_into_ellipsis(&self,
+                                   layout_context: &LayoutContext,
+                                   text_overflow_string: String)
+                                   -> Fragment {
         let mut unscanned_ellipsis_fragments = LinkedList::new();
         unscanned_ellipsis_fragments.push_back(self.transform(
                 self.border_box.size,
                 SpecificFragmentInfo::UnscannedText(
-                    box UnscannedTextFragmentInfo::new("…".to_owned(), None))));
+                    box UnscannedTextFragmentInfo::new(text_overflow_string, None))));
         let ellipsis_fragments = TextRunScanner::new().scan_for_runs(&mut layout_context.font_context(),
                                                                      unscanned_ellipsis_fragments);
         debug_assert!(ellipsis_fragments.len() == 1);

--- a/components/layout/inline.rs
+++ b/components/layout/inline.rs
@@ -31,10 +31,10 @@ use std::collections::VecDeque;
 use std::sync::Arc;
 use style::arc_ptr_eq;
 use style::computed_values::{display, overflow_x, position, text_align, text_justify};
-use style::computed_values::{text_overflow, vertical_align, white_space};
+use style::computed_values::{vertical_align, white_space};
 use style::context::{SharedStyleContext, StyleContext};
 use style::logical_geometry::{LogicalRect, LogicalSize, WritingMode};
-use style::properties::ServoComputedValues;
+use style::properties::{longhands, ServoComputedValues};
 use style::servo::restyle_damage::{BUBBLE_ISIZES, REFLOW, REFLOW_OUT_OF_FLOW, REPOSITION, RESOLVE_GENERATED_CONTENT};
 use text;
 use unicode_bidi;
@@ -684,21 +684,30 @@ impl LineBreaker {
         }
 
         // Determine if an ellipsis will be necessary to account for `text-overflow`.
-        let mut need_ellipsis = false;
         let available_inline_size = self.pending_line.green_zone.inline -
             self.pending_line.bounds.size.inline - indentation;
-        match (fragment.style().get_text().text_overflow,
-               fragment.style().get_box().overflow_x) {
-            (text_overflow::T::clip, _) | (_, overflow_x::T::visible) => {}
-            (text_overflow::T::ellipsis, _) => {
-                need_ellipsis = fragment.margin_box_inline_size() > available_inline_size;
-            }
-        }
 
-        if !need_ellipsis {
-            self.push_fragment_to_line_ignoring_text_overflow(fragment, layout_context);
-        } else {
-            let ellipsis = fragment.transform_into_ellipsis(layout_context);
+        let ellipsis = match (&fragment.style().get_text().text_overflow.first,
+            fragment.style().get_box().overflow_x) {
+            (&longhands::text_overflow::Side::Clip, _) | (_, overflow_x::T::visible) => None,
+            (&longhands::text_overflow::Side::Ellipsis, _) => {
+                if fragment.margin_box_inline_size() > available_inline_size {
+                    Some("…".to_string())
+                } else {
+                    None
+                }
+            },
+            (&longhands::text_overflow::Side::String(ref string), _) => {
+                if fragment.margin_box_inline_size() > available_inline_size {
+                    Some(string.to_string())
+                } else {
+                    None
+                }
+            }
+        };
+
+        if let Some(string) = ellipsis {
+            let ellipsis = fragment.transform_into_ellipsis(layout_context, string);
             if let Some(truncation_info) =
                     fragment.truncate_to_inline_size(available_inline_size -
                                                      ellipsis.margin_box_inline_size()) {
@@ -707,6 +716,8 @@ impl LineBreaker {
                 self.push_fragment_to_line_ignoring_text_overflow(fragment, layout_context);
             }
             self.push_fragment_to_line_ignoring_text_overflow(ellipsis, layout_context);
+        } else {
+            self.push_fragment_to_line_ignoring_text_overflow(fragment, layout_context);
         }
 
         if line_flush_mode == LineFlushMode::Flush {

--- a/components/style/properties/longhand/text.mako.rs
+++ b/components/style/properties/longhand/text.mako.rs
@@ -12,9 +12,6 @@
                                              Method("has_overline", "bool"),
                                              Method("has_line_through", "bool")]) %>
 
-% if product == "servo":
-    ${helpers.single_keyword("text-overflow", "clip ellipsis", animatable=False)}
-% else:
 <%helpers:longhand name="text-overflow" animatable="False">
     use std::fmt;
     use style_traits::ToCss;
@@ -93,7 +90,6 @@
         }
     }
 </%helpers:longhand>
-% endif
 
 ${helpers.single_keyword("unicode-bidi",
                          "normal embed isolate bidi-override isolate-override plaintext",

--- a/tests/wpt/mozilla/meta/MANIFEST.json
+++ b/tests/wpt/mozilla/meta/MANIFEST.json
@@ -5496,18 +5496,6 @@
             "url": "/_mozilla/css/text_node_opacity.html"
           }
         ],
-        "css/text_overflow_a.html": [
-          {
-            "path": "css/text_overflow_a.html",
-            "references": [
-              [
-                "/_mozilla/css/text_overflow_ref.html",
-                "!="
-              ]
-            ],
-            "url": "/_mozilla/css/text_overflow_a.html"
-          }
-        ],
         "css/text_overflow_basic_a.html": [
           {
             "path": "css/text_overflow_basic_a.html",
@@ -5518,6 +5506,30 @@
               ]
             ],
             "url": "/_mozilla/css/text_overflow_basic_a.html"
+          }
+        ],
+        "css/text_overflow_ellipsis.html": [
+          {
+            "path": "css/text_overflow_ellipsis.html",
+            "references": [
+              [
+                "/_mozilla/css/text_overflow_ref.html",
+                "!="
+              ]
+            ],
+            "url": "/_mozilla/css/text_overflow_ellipsis.html"
+          }
+        ],
+        "css/text_overflow_string.html": [
+          {
+            "path": "css/text_overflow_string.html",
+            "references": [
+              [
+                "/_mozilla/css/text_overflow_ref.html",
+                "!="
+              ]
+            ],
+            "url": "/_mozilla/css/text_overflow_string.html"
           }
         ],
         "css/text_shadow_blur_a.html": [
@@ -20364,18 +20376,6 @@
           "url": "/_mozilla/css/text_node_opacity.html"
         }
       ],
-      "css/text_overflow_a.html": [
-        {
-          "path": "css/text_overflow_a.html",
-          "references": [
-            [
-              "/_mozilla/css/text_overflow_ref.html",
-              "!="
-            ]
-          ],
-          "url": "/_mozilla/css/text_overflow_a.html"
-        }
-      ],
       "css/text_overflow_basic_a.html": [
         {
           "path": "css/text_overflow_basic_a.html",
@@ -20386,6 +20386,30 @@
             ]
           ],
           "url": "/_mozilla/css/text_overflow_basic_a.html"
+        }
+      ],
+      "css/text_overflow_ellipsis.html": [
+        {
+          "path": "css/text_overflow_ellipsis.html",
+          "references": [
+            [
+              "/_mozilla/css/text_overflow_ref.html",
+              "!="
+            ]
+          ],
+          "url": "/_mozilla/css/text_overflow_ellipsis.html"
+        }
+      ],
+      "css/text_overflow_string.html": [
+        {
+          "path": "css/text_overflow_string.html",
+          "references": [
+            [
+              "/_mozilla/css/text_overflow_ref.html",
+              "!="
+            ]
+          ],
+          "url": "/_mozilla/css/text_overflow_string.html"
         }
       ],
       "css/text_shadow_blur_a.html": [

--- a/tests/wpt/mozilla/tests/css/text_overflow_ellipsis.html
+++ b/tests/wpt/mozilla/tests/css/text_overflow_ellipsis.html
@@ -1,0 +1,20 @@
+<!DOCTYPE html>
+<html>
+<head>
+<link rel=mismatch href=text_overflow_ref.html>
+<style>
+p {
+    width: 128px;
+    background: gold;
+    text-overflow: ellipsis;
+}
+p.hidden {
+    overflow: hidden;
+}
+</style>
+</head>
+<body>
+<p>mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
+<p class="hidden">mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
+</body>
+</html>

--- a/tests/wpt/mozilla/tests/css/text_overflow_ref.html
+++ b/tests/wpt/mozilla/tests/css/text_overflow_ref.html
@@ -13,5 +13,3 @@ p {
 <p>mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
 </body>
 </html>
-
-

--- a/tests/wpt/mozilla/tests/css/text_overflow_string.html
+++ b/tests/wpt/mozilla/tests/css/text_overflow_string.html
@@ -6,16 +6,15 @@
 p {
     width: 128px;
     background: gold;
-    text-overflow: ellipsis;
+    text-overflow: "[..]";
 }
-#hideme {
+p.hidden {
     overflow: hidden;
 }
 </style>
 </head>
 <body>
-<p id=hideme>mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
 <p>mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
+<p class="hidden">mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm</p>
 </body>
 </html>
-

